### PR TITLE
Fixes #807 fails if unit is provided and path is wrong

### LINKED ISSUE #807: `setParameterValuesByPath` fails if the path cannot be found if a unit is provided, even if `stopIfNotFound = FALSE`
Test: 

```
  sim <- loadTestSimulation("S1", loadFromCache = TRUE)
  parameterPath <- "foo"
  setParameterValuesByPath(parameterPath, 100, sim, stopIfNotFound = FALSE, units = "l")

Error in rClr::clrCall(task, "DimensionNameByPath", simulation$ref, path) : 
Type: OSPSuite.Utility.Exceptions.OSPSuiteException
Message: Could not find quantity with path 'foo'.
Method: OSPSuite.Core.Domain.IQuantity singleQuantityByPath(OSPSuite.Core.Domain.IModelCoreSimulation, System.String)
Stack trace:
at OSPSuite.R.Services.ContainerTask.singleQuantityByPath(IModelCoreSimulation simulation, String path)
at OSPSuite.R.Services.ContainerTask.DimensionNameByPath(IModelCoreSimulation simulation, String path)
3.
rClr::clrCall(task, "DimensionNameByPath", simulation$ref, path) at utilities-quantity.R#158
2.
setQuantityValuesByPath(quantityPaths = parameterPaths, values = values, 
simulation = simulation, units = units, stopIfNotFound = stopIfNotFound) at utilities-parameter.R#128
1.
setParameterValuesByPath(parameterPath, 100, sim, stopIfNotFound = FALSE, 
units = "l")
```

### PR DESCRIPTION
**Pull request by msevestre**

### CODE REVIEW

File: tests/testthat/test-utilities-quantity.R [hunk 1]
```diff
@@ -56,7 +56,7 @@ test_that("It returns null if the quantity by path does not exist and stopIfNotF
 })
 
 test_that("It throws an error when trying to retrieve a quantity by path that would result in multiple quantities", {
-  expect_that(getQuantity(toPathString(c("Organism", "Liver", "*")), sim), throws_error())
+  expect_error(getQuantity(toPathString(c("Organism", "Liver", "*")), sim))
```
Comment (IndrajeetPatil) [inline tests/testthat/test-utilities-quantity.R:59]: I think we should also be checking the exact error thrown here using the `regexp` [argument](https://testthat.r-lib.org/reference/expect_error.html).

This way, we know that the expected error occurred.

File: tests/testthat/test-utilities-quantity.R [hunk 2]
```diff
@@ -102,17 +102,21 @@ test_that("It can set multiple quantity values with units", {
 
 test_that("It throws an exception when setting values for a quantity that does not exist", {
   parameterPath <- "Organism|Liver|NOPE|Volume"
-  expect_that(setQuantityValuesByPath(parameterPath, 100, sim), throws_error())
+  expect_error(setQuantityValuesByPath(parameterPath, 100, sim))
+})
+
+test_that("It does not throw an exception when setting values for a quantity that does not exist with unit", {
+  parameterPath <- "Organism|Liver|NOPE|Volume"
+  expect_error(setQuantityValuesByPath(quantityPaths = parameterPath, values = 100, simulation = sim, units = "ml", stopIfNotFound = FALSE), NA)
```
Comment (IndrajeetPatil) [inline tests/testthat/test-utilities-quantity.R:110]: See above. 

The same for the two other instances below of `expect_error()`.

File: R/utilities-simulation-results.R [hunk 3]
```diff
@@ -27,8 +27,6 @@
 #' @param population population used to calculate the `simulationResults`
 #'   (optional). This is used only to add the population covariates to the
 #'   resulting data table.
-#' @param stopIfNotFound Boolean. If `TRUE` and no result exist for the given
```
Comment (IndrajeetPatil) [inline R/utilities-simulation-results.R:30]: DRY docs. Nice!

**Review by IndrajeetPatil — CHANGES_REQUESTED** [review]

File: R/utilities-quantity.R [hunk 4]
```diff
@@ -155,8 +155,11 @@ setQuantityValuesByPath <- function(quantityPaths, values, simulation, units = N
     path <- enc2utf8(quantityPaths[[i]])
     value <- values[[i]]
     if (!is.null(units)) {
-      dimension <- rClr::clrCall(task, "DimensionNameByPath", simulation$ref, path)
-      value <- toBaseUnit(quantityOrDimension = dimension, values = value, unit = units[[i]])
+      dimension <- rClr::clrCall(task, "DimensionNameByPath", simulation$ref, path, stopIfNotFound)
+      # Dimension ca be be empty if the path was not found and stopIfNotFound is false
+      if (dimension != "" || stopIfNotFound) {
```
Comment (PavelBal) [inline R/utilities-quantity.R:160]: I am still not quite sure I understand this :D

I think you should exit early, if dimension is `""`, than the path is definitely not found and no need to continue:

```
    if (!is.null(units)) {
      dimension <- rClr::clrCall(task, "DimensionNameByPath", simulation$ref, path, stopIfNotFound)
      # Dimension ca be be empty if the path was not found and stopIfNotFound is false
      if (dimension == "") {
        next
      }
      value <- toBaseUnit(quantityOrDimension = dimension, values = value, unit = units[[i]])
    }
```

File: tests/testthat/test-utilities-quantity.R (hunk 1 above)
Comment (PavelBal) [reply tests/testthat/test-utilities-quantity.R:59]: Yes, I agree, we should try to follow this practice. Sometimes the error is thrown and the test passes, but the error is actually somewhere else and we do not really test the case.

File: tests/testthat/test-utilities-quantity.R (hunk 2 above)
Comment (PavelBal) [reply tests/testthat/test-utilities-quantity.R:110]: Here we are actually testing that it does NOT throw an error :D

Comment (IndrajeetPatil) [reply tests/testthat/test-utilities-quantity.R:110]: Ah, I think, then this should be `expect_output()`.

File: tests/testthat/test-utilities-quantity.R (hunk 1 above)
Comment (PavelBal) [reply tests/testthat/test-utilities-quantity.R]: I added the message

File: tests/testthat/test-utilities-quantity.R (hunk 2 above)
Comment (PavelBal) [reply tests/testthat/test-utilities-quantity.R]: Changed to expect_output()

File: tests/testthat/test-utilities-quantity.R [hunk 5]
```diff
@@ -102,17 +103,41 @@ test_that("It can set multiple quantity values with units", {
...
 
 test_that("It throws an error when the number of quantity paths differs from the number units", {
   quantityPath1 <- "Organism|Liver|Intracellular|Volume"
   quantityPath2 <- "Organism|VenousBlood|Plasma|CYP3A4"
-  expect_that(setQuantityValuesByPath(c(quantityPath1, quantityPath2), c(40, 50), sim, units = "ml"), throws_error())
+  expect_error(setQuantityValuesByPath(quantityPaths = c(quantityPath1, quantityPath2), values = c(40, 50),simulation =  sim, units = "ml"),
+               regexp = messages$errorDifferentLength("quantityPaths, units"))
+})
+
+test_that("It throws an error when setting values for a set of quantities with units and one path does not exist", {
```
Comment (PavelBal) [inline tests/testthat/test-utilities-quantity.R:126]: I added two tests for multiple paths.

**Comment by PavelBal** [comment]
@msevestre Thanks for fixing it so quickly, amazing. I just have one suggestion to exit early when the dimension could not be retrieved. However, this does not throw a warning "path could not be found", this should be extended in .NET maybe? (showing a warning when entity cannot be found).

**Review by IndrajeetPatil — APPROVED** [review]

File: tests/testthat/test-utilities-quantity.R (hunk 2 above)
Comment (msevestre) [reply tests/testthat/test-utilities-quantity.R]: `expect_output` !! I was wondering how to test that there is NO Error lol. THANK YOU @IndrajeetPatil . A quick search in SO showed that I need to put NA at the end. But this looked very weird to me and I was about to write an expect_no_error method lol (which would be easier to understand than expect_output IMO)

File: tests/testthat/test-utilities-quantity.R (hunk 1 above)
Comment (msevestre) [reply tests/testthat/test-utilities-quantity.R]: Interesting. I see pros and cons. Let's discuss

File: R/utilities-quantity.R (hunk 4 above)
Comment (msevestre) [reply R/utilities-quantity.R]: done

**Comment by msevestre** [comment]
ok warning also when dimension or unit is not  found.
Once the build is running, I'll nerge

**Pull request merged**